src/v0/destinations/google_adwords_offline_conversions/utils.js

```diff
@@ -465,6 +550,269 @@ const getConsentsDataFromIntegrationObj = (message) => {
+  };
+  const endpoint = SEARCH_STREAM.replace(CUSTOMER_ID_PARAM, customerId);
+  const requestOptions = {
+    headers,
+  };
+
+  let searchStreamResponse = await httpPOST(endpoint, data, requestOptions, {
+    destType: 'google_adwords_offline_conversions',
+    feature: 'transformation',
+    endpointPath: `/googleAds:searchStream`,
+    requestMethod: 'POST',
+    module: 'transformation',
+    metadata,
+  });
+  searchStreamResponse = processAxiosResponse(searchStreamResponse);
+  const { response, status } = searchStreamResponse;
+
+  if (!isHttpStatusSuccess(status)) {
+    throw new NetworkError(
+      `[Google Ads Offline Conversions]:: ${response?.[0]?.error?.message || response?.error?.message} during batch conversion custom variable fetch`,
+      status,
+      {
+        [tags.TAG_NAMES.ERROR_TYPE]: getDynamicErrorType(status),
+      },
+      response || searchStreamResponse,
+      getAuthErrCategory(searchStreamResponse),
+    );
+  }
+
+  const results = get(searchStreamResponse, 'response.0.results') || [];
+  const variableMap = {};
+
+  // Build result map from API response
+  for (const resultItem of results) {
+    const variable = resultItem.conversionCustomVariable;
+    if (variable?.name && variable?.resourceName) {
+      variableMap[variable.name] = variable.resourceName;
+    }
+  }
+
+  return variableMap;
+};
+
+/**
+ * Get conversion custom variables for a customer ID with batch fetching support
+ * First checks cache for each variable individually. If any variable is missing,
+ * fetches ALL variables in a single API call (since the API call cost is the same
+ * whether fetching one or many custom variables).
+ * Uses individual cache keys: customerId + variableName
+ * @param {string} customerId - The customer ID
+ * @param {string[]} variableNames - Array of custom variable names needed
+ * @param {object} headers - Request headers
+ * @param {object} metadata - Request metadata
+ * @returns {Object} Map of variable names to resource names
+ */
+const getConversionCustomVariables = async ({ Config, metadata, customerId, variableNames }) => {
+  if (!Array.isArray(variableNames) || variableNames.length === 0) {
+    return {};
+  }
+
+  const result = {};
+  let hasCacheMiss = false;
+
+  // Check cache for each variable individually (synchronous check, no await in loop)
+  for (const variableName of variableNames) {
+    const cacheKey = sha256(customerId + variableName).toString();
+    // Access cache directly without storeFunction
+    // eslint-disable-next-line no-await-in-loop
+    const cachedValue = await conversionCustomVariableCache.get(cacheKey);
+
+    if (cachedValue) {
+      result[variableName] = cachedValue;
+    } else {
+      hasCacheMiss = true;
+    }
+  }
+
+  // If there are any cache misses, fetch ALL variables in single API call
+  // since the API call cost is the same regardless of how many we fetch
+  if (hasCacheMiss) {
+    const headers = getHeader(Config, metadata, true);
```

Comment:
Headers resolution responsibility should be moved to the `batchFetch...` function as it makes the request.